controller/server_controller.py

```diff
@@ -288,78 +335,127 @@ def cull_idle_jupyter_servers(body, name, namespace, logger, **kwargs):
+            action="hibernate",
+            name=name,
+            namespace=namespace,
+            logger=logger,
+            body={
+                "metadata": {
+                    "annotations": {
+                        "renku.io/hibernation": json.dumps(hibernation),
+                        "renku.io/hibernation-branch": "",
+                        "renku.io/hibernation-commit-sha": "",
+                        "renku.io/hibernation-dirty": "",
+                        "renku.io/hibernation-synchronized": "",
+                        "renku.io/hibernation-date": now,
+                    },
+                },
+                "spec": {
+                    "jupyterServer": {
+                        "hibernated": True,
+                    },
+                },
+            },
+        )
         return
 
     if jupyter_server_is_idle_now:
+        idle_seconds += config.JUPYTER_SERVER_IDLE_CHECK_INTERVAL_SECONDS
         logger.info(
-            f"Jupyter Server {name} in namespace {namespace} found to be idle for "
-            f"{idle_seconds + config.JUPYTER_SERVER_IDLE_CHECK_INTERVAL_SECONDS}"
+            f"Jupyter Server {name} in namespace {namespace} found to be idle for {idle_seconds}"
+        )
+        update_idle_seconds(idle_seconds)
+    elif idle_seconds > 0:
+        logger.info(
+            f"Resetting idle timer for Jupyter server {name} in namespace {namespace}."
+        )
+        update_idle_seconds(0, action="reset")
+
+
+@kopf.timer(
+    config.api_group,
+    config.api_version,
+    config.custom_resource_name,
+    interval=config.JUPYTER_SERVER_IDLE_CHECK_INTERVAL_SECONDS,
+)
+def cull_hibernated_jupyter_servers(body, name, namespace, logger, **_):
+    """Check if a server is hibernated for long enough, then cull it."""
+
+    def update_hibernated_seconds(seconds, action="update"):
+        patch_jupyter_servers(
+            action=f"{action} hibernated timer",
+            name=name,
+            namespace=namespace,
+            logger=logger,
+            body={
+                "status": {
+                    "hibernatedSeconds": str(seconds),
+                },
+            },
+        )
+
+    hibernated = body.get("spec", {}).get("jupyterServer", {}).get("hibernated")
+    hibernated_seconds = int(body.get("status", {}).get("hibernatedSeconds", "0"))
+
+    # NOTE: Do nothing if the server isn't hibernated
+    if not hibernated:
+        if hibernated_seconds:
+            update_hibernated_seconds(0, action="reset")
+        return
+
+    hibernated_seconds_threshold = body["spec"]["culling"].get("hibernatedSecondsThreshold", 0)
+    if hibernated_seconds_threshold == 0:
+        return
+
+    annotations = body.get("metadata", {}).get("annotations", {})
+    hibernation_date_str = annotations.get("renku.io/hibernation-date", "")
+
+    # NOTE: ``hibernation-date`` is ``""`` when session isn't hibernated
+    if not hibernation_date_str:
+        return
```

Comment:
This case should never happen since we patch both hibernation date and hibernation flag at the same time. If in some imaginary case this is happening, then something is very wrong with the system and I prefer to preserve the session instead of deleting it. Do you agree?

Reply:
This case can definitely happen when an admin edits the spec manually through kubectl. In this case we should not expect the admin to also add the hibernation date manually.

Reply:
I ignored this case. Fixed.